### PR TITLE
ISPN-8984 Dynamic reversioning

### PR DESCRIPTION
**Pull request by tristantarrant**
ISPN-8951 LevelDB parser stopped at 8.0 schema
ISPN-8952 Support wildcard parsers
ISPN-8953 Remove old schemas
ISPN-8984 Reversion script now updates XSDs
ISPN-8985 Dynamically use latest schema version in tests

Also normalize pom module names

https://issues.jboss.org/browse/ISPN-8951
https://issues.jboss.org/browse/ISPN-8952
https://issues.jboss.org/browse/ISPN-8953
https://issues.jboss.org/browse/ISPN-8984
https://issues.jboss.org/browse/ISPN-8985

This needs to land before we branch for 9.3

### CODE REVIEW

**Comment by gustavonalle** [comment]
@tristantarrant There are 100+ failures, could you check?

File: bin/reversion.sh [hunk 1]
```diff
@@ -1,16 +1,40 @@
...
+        # Rewrite the filename
+        NEWSCHEMA=`echo "$OLDSCHEMA"|sed "s/$OLDMAJOR$SEPARATOR$OLDMINOR/$NEWMAJOR$SEPARATOR$NEWMINOR/"`
+        echo "$OLDSCHEMA ==> $NEWSCHEMA"
+        mv $OLDSCHEMA $NEWSCHEMA
+        # We need to do different replacements
+        sed -i "s/:$OLDMAJOR.$OLDMINOR\"/:$NEWMAJOR.$NEWMINOR\"/g" $NEWSCHEMA
+        sed -i "s/-$OLDMAJOR$SEPARATOR$OLDMINOR.xsd/-$NEWMAJOR$SEPARATOR$NEWMINOR.xsd/g" $NEWSCHEMA
+        git rm $OLDSCHEMA
+        git add $NEWSCHEMA
+    done
```
Comment (gustavonalle) [inline bin/reversion.sh]: There are many occurrences which are not being replaced:
```grep '9.2' **/*.xml``` (zsh)

**Comment by tristantarrant** [comment]
Yes, last minute search/replace screw up

File: bin/reversion.sh [hunk 2]
```diff
@@ -43,4 +71,27 @@ fi
 
 echo "Changing from $VERSION to $NEWVERSION"
 
-find . -name 'pom.xml' -exec sed -i "s/<version>$VERSION<\/version>/<version>$NEWVERSION<\/version>/g" {} \;
+find . -name 'pom.xml' -exec sed -i "s/<version>$VERSION<\/version>/<version>$NEWVERSION<\/version>/g" {} \; -exec git add {} \;
```
Comment (gustavonalle) [inline bin/reversion.sh]: The documentations's ```pom.xml```  still declares ```<infinispan.version>9.2</infinispan.version>``` after replacement

Comment (gustavonalle) [reply bin/reversion.sh]: In ```server/integration/testsuite/pom.xml``` there is still ```<new.server.schema.version>9.2</new.server.schema.version>```

Comment (gustavonalle) [reply bin/reversion.sh]: In ```server/integration/versions/pom.xml``` there is  
```<version.org.infinispan.infinispan-management-console>9.2.0.Final</version.org.infinispan.infinispan-management-console>```

Which is fine, just a reminder that this should not be replaced :smile:

File: bin/reversion.sh [hunk 3]
```diff
@@ -43,4 +71,27 @@ fi
...
+NEWSCHEMAMAJOR=`echo $NEWVERSION|cut -d. -f1`
+NEWSCHEMAMINOR=`echo $NEWVERSION|cut -d. -f2`
+NEWSCHEMAVERSION=`echo $NEWVERSION|cut -d. -f1,2`
+
+if [ "$OLDSCHEMAVERSION" != "$NEWSCHEMAVERSION" ] && [ "$PROCESS_SCHEMAS" = true ] ;  then
+    echo "Current schema: $OLDSCHEMAVERSION"
+    echo "New schema:     $NEWSCHEMAVERSION"
+    # Create new unified configuration
+    cp core/src/test/resources/configs/unified/$OLDSCHEMAVERSION.xml core/src/test/resources/configs/unified/$NEWSCHEMAVERSION.xml
+    sed -i "s/:$OLDSCHEMAMAJOR.$OLDSCHEMAMINOR/:$NEWSCHEMAMAJOR.$NEWSCHEMAMINOR/g" core/src/test/resources/configs/unified/$NEWSCHEMAVERSION.xml
```
Comment (gustavonalle) [inline bin/reversion.sh]: This file still contains ``` xsi:schemaLocation="urn:infinispan:config:9.3 http://www.infinispan.org/schemas/infinispan-config-9.2.xsd"```

Comment (gustavonalle) [reply bin/reversion.sh]: (after replacement)

Comment (gustavonalle) [reply bin/reversion.sh]: There is maven goal to extract version components and expose them as variables, this could be used for those major.minor strings
http://www.mojohaus.org/build-helper-maven-plugin/parse-version-mojo.html

File: bin/reversion.sh (hunk 2 above)
Comment (tristantarrant) [reply bin/reversion.sh]: The documentation property can be removed as it is ignored.

**Comment by tristantarrant** [comment]
@gustavonalle the latest CI build is fine. I should have removed all instances of hardcoded schema versions (outside of those required by rolling upgrades et al).
The reversion script still requires some manual intervention in the server Namespace classes. Not sure how to fix this aside from using some kind of source templating engine.

**Comment by gustavonalle** [comment]
looking

File: spring/spring4/spring4-remote/src/test/resources/org/infinispan/spring/provider/named-async-cache.xml [hunk 4]
```diff
@@ -1,8 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (gustavonalle) [inline spring/spring4/spring4-remote/src/test/resources/org/infinispan/spring/provider/named-async-cache.xml]: Duplicate line

File: server/integration/jgroups/src/test/java/org/infinispan/server/jgroups/subsystem/SubsystemParsingTestCase.java [hunk 5]
```diff
@@ -71,11 +71,15 @@ public SubsystemParsingTestCase(JGroupsSchema schema, int expectedOperationCount
...
-                { JGroupsSchema.INFINISPAN_SERVER_JGROUPS_9_2, 27, "schema/jboss-infinispan-jgroups_9_2.xsd", new String[] { "/subsystem-templates/infinispan-jgroups.xml", "/subsystem-templates/cloud-jgroups.xml" } },
+        Object[][] data = new Object[][]{
+              {JGroupsSchema.INFINISPAN_SERVER_JGROUPS_7_0, 25, null, null},
+              {JGroupsSchema.INFINISPAN_SERVER_JGROUPS_8_0, 27, null, null},
+              {JGroupsSchema.INFINISPAN_SERVER_JGROUPS_9_0, 27, null, null},
+              {
+                    JGroupsSchema.INFINISPAN_SERVER_JGROUPS_9_2, 27,
+                    "schema/jboss-infinispan-jgroups_9_2.xsd",
+                    new String[]{"/subsystem-templates/infinispan-jgroups.xml", "/subsystem-templates/cloud-jgroups.xml"}
+              },
```
Comment (gustavonalle) [inline server/integration/jgroups/src/test/java/org/infinispan/server/jgroups/subsystem/SubsystemParsingTestCase.java]: Weird alignment :)

File: lucene/directory-provider/src/test/resources/testing-flush-loopback.xml [hunk 6]
```diff
@@ -28,7 +28,7 @@
 	<pbcast.STABLE stability_delay="1000" desired_avg_gossip="50000"
 		max_bytes="400000" />
 	<pbcast.GMS
-			print_local_addr="true"
+			print_local_addr="false"
```
Comment (gustavonalle) [inline lucene/directory-provider/src/test/resources/testing-flush-loopback.xml]: Is this change related?

File: bin/reversion.sh [hunk 7]
```diff
@@ -1,16 +1,51 @@
 #!/bin/bash
```
Comment (gustavonalle) [inline bin/reversion.sh]: I am getting errors when running this script on macOS High Sierra

```
➜  infinispan git:(ISPN-8984/reversioning) bin/reversion.sh 
readlink: illegal option -- f
usage: readlink [-n] [file ...]
warning: failed to load external entity "/extract-gav.xslt"
cannot parse /extract-gav.xslt
Pom file location: pom.xml
Current version: 
New version: 9.3.0-SNAPSHOT
Changing from  to 9.3.0-SNAPSHOT
sed: 1: "./tree/pom.xml": invalid command code .
sed: 1: "./documentation/pom.xml": invalid command code .
sed: 1: "./checkstyle/pom.xml": invalid command code .
```

**Comment by gustavonalle** [comment]
I am getting errors when running this script on macOS High Sierra

```
➜  infinispan git:(ISPN-8984/reversioning) bin/reversion.sh 
readlink: illegal option -- f
usage: readlink [-n] [file ...]
warning: failed to load external entity "/extract-gav.xslt"
cannot parse /extract-gav.xslt
Pom file location: pom.xml
Current version: 
New version: 9.3.0-SNAPSHOT
Changing from  to 9.3.0-SNAPSHOT
sed: 1: "./tree/pom.xml": invalid command code .
sed: 1: "./documentation/pom.xml": invalid command code .
sed: 1: "./checkstyle/pom.xml": invalid command code .
```


On Linux (Fedora) it seems to work fine. I think we can fix the MacOS later. The fix would be to port this script to python 🐍 ;)

**Comment by gustavonalle** [comment]
What would be the manual required after running the scripts? 
Could you document in somewhere in the code?

File: spring/spring4/spring4-remote/src/test/resources/org/infinispan/spring/provider/named-async-cache.xml (hunk 4 above)
Comment (tristantarrant) [reply spring/spring4/spring4-remote/src/test/resources/org/infinispan/spring/provider/named-async-cache.xml]: Ok

File: server/integration/jgroups/src/test/java/org/infinispan/server/jgroups/subsystem/SubsystemParsingTestCase.java (hunk 5 above)
Comment (tristantarrant) [reply server/integration/jgroups/src/test/java/org/infinispan/server/jgroups/subsystem/SubsystemParsingTestCase.java]: Indeed. That is what IDEA gives me. I'll hand-format it.

File: lucene/directory-provider/src/test/resources/testing-flush-loopback.xml (hunk 6 above)
Comment (tristantarrant) [reply lucene/directory-provider/src/test/resources/testing-flush-loopback.xml]: Not really, but it merely aligns it with the behaviour of all other tests. Harmless.

File: bin/reversion.sh (hunk 2 above)
Comment (danberindei) [reply bin/reversion.sh]: I'd rather split this over multiple lines:

```
for f in $(find . ...); do
  sed -i ...
  git add ...
done
```

**Comment by gustavonalle** [comment]
@tristantarrant LGTM. I created https://issues.jboss.org/browse/ISPN-9012 to support macOS. 
BTW, there is one suggestion by @danberindei unadressed, but if we are going to do ISPN-9012, I think it won't matter much since the bash script will be rewritten

**Comment by tristantarrant** [comment]
Which suggestion ? I don't see any comments by @danberindei  here

**Comment by gustavonalle** [comment]
@tristantarrant @danberindei Looks like the comment disappeared or was deleted...But I can see in my email

**Comment by tristantarrant** [comment]
Ah, right, I saw it in my e-mails too, addressed it and repushed, but I thought it was from you :)

**Comment by gustavonalle** [comment]
Merged, thanks @tristantarrant !
